big_vision/trainers/proj/gsam/train.py

```diff
@@ -0,0 +1,401 @@
+       "%d dev per host (%d dev total), that's a %d per-device batch size.",
+       batch_size, jax.process_count(), batch_size // jax.process_count(),
+       jax.local_device_count(), jax.device_count(),
+       batch_size // jax.device_count())
+
+  # First thing after above sanity checks, so we can log "start" ticks.
+  mw = u.BigVisionMetricWriter(xid, wid, workdir)
+  chrono = u.Chrono()
+
+  write_note("Initializing train dataset...")
+  train_ds = input_pipeline.make_for_train(
+      dataset=config.dataset,
+      split=config.train_split,
+      batch_size=config.batch_size,
+      preprocess_fn=pp_builder.get_preprocess_fn(config.pp_train),
+      shuffle_buffer_size=config.get("shuffle_buffer_size"),
+      cache_raw=config.get("cache_raw", False),
+      data_dir=fillin(config.get("dataset_dir")))
+
+  n_prefetch = config.get("prefetch_to_device", 1)
+  train_iter = input_pipeline.start_input_pipeline(train_ds, n_prefetch)
+
+  ntrain_img = input_pipeline.get_num_examples(
+      config.dataset, config.train_split,
+      data_dir=fillin(config.get("dataset_dir")))
+  steps_per_epoch = ntrain_img / batch_size
+
+  if config.get("num_epochs"):
+    total_steps = int(config.num_epochs * steps_per_epoch)
+    assert not config.get("total_steps"), "Set either num_epochs or total_steps"
+  else:
+    total_steps = config.total_steps
+
+  info("Running for %d steps, that means %f epochs and %f steps per epoch",
+       total_steps, total_steps * batch_size / ntrain_img, steps_per_epoch)
+
+  write_note(f"Initializing {config.model_name} model...")
+  model_mod = importlib.import_module(f"big_vision.models.{config.model_name}")
+  model = model_mod.Model(
+      num_classes=config.num_classes, **config.get("model", {}))
+
+  # We want all parameters to be created in host RAM, not on any device, they'll
+  # be sent there later as needed, otherwise we already encountered two
+  # situations where we allocate them twice.
+  @partial(jax.jit, backend="cpu")
+  def init(rng):
+    shape = tuple(train_ds.element_spec["image"].shape[1:])
+    bs = config.batch_size // jax.device_count()
+    dummy_input = jnp.zeros((bs,) + shape, jnp.float32)
+    params = flax.core.unfreeze(model.init(rng, dummy_input))["params"]
+
+    # Set bias in the head to a low value, such that loss is small initially.
+    if "init_head_bias" in config:
+      params["head"]["bias"] = jnp.full_like(params["head"]["bias"],
+                                             config["init_head_bias"])
+
+    return params
+
+  rng, rng_init = jax.random.split(rng)
+  params_cpu = init(rng_init)
+
+  if jax.process_index() == 0:
+    num_params = sum(p.size for p in jax.tree_leaves(params_cpu))
+    parameter_overview.log_parameter_overview(params_cpu, msg="init params")
+    mw.measure("num_params", num_params)
+
+  write_note(f"Initializing {config.optax_name} optimizer...")
+  tx, sched_fns = bv_optax.make(config, params_cpu, sched_kw=dict(
```

Comment:
Given that the way you implement GSAM it only supports a single global learning-rate, could you add something here similar to:

assert len(sched_fns) == 1, "Current GSAM implementation only supports one global learning-rate"

Reply:
Done